timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/HistoryAnalyzer.java

```diff
@@ -0,0 +1,143 @@
+
+    private static boolean learnedValueIsGreatestAcceptedValue(
+            List<ConsolidatedLearnerAndAcceptorRecord> records, Long seq) {
+        byte[] learnedValueData = getPaxosValueData(getLearnedValue(records, seq));
+        if (learnedValueData == null) {
+            return true;
+        }
+
+        byte[] greatestAcceptedValueData = getPaxosValueData(getGreatestAcceptedValueAtSequence(records, seq));
+        if (greatestAcceptedValueData == null) {
+            // should not reach here
+            return false;
+        }
+        return PtBytes.toLong(greatestAcceptedValueData) <= PtBytes.toLong(learnedValueData);
+    }
+
+    private static Optional<PaxosValue> getLearnedValue(List<ConsolidatedLearnerAndAcceptorRecord> recordList, Long seq) {
+        Set<PaxosValue> values = getLearnedValuesForRound(recordList, seq);
+        return values.isEmpty() ? Optional.empty() : Optional.of(Iterables.getOnlyElement(values));
+    }
+
+    private static Set<PaxosValue> getLearnedValuesForRound(List<ConsolidatedLearnerAndAcceptorRecord> recordList,
+            Long seq) {
+        return recordList.stream()
+                .map(consolidatedLearnerAndAcceptorRecord ->
+                        consolidatedLearnerAndAcceptorRecord.get(seq).learnedValue())
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.toSet());
+    }
+
+    private static int getQuorumSize(List<ConsolidatedLearnerAndAcceptorRecord> records) {
+        return records.size() / 2 + 1;
+    }
+
+    private static List<PaxosValue> getAcceptedValues(List<ConsolidatedLearnerAndAcceptorRecord> records, Long seq,
+            PaxosValue learnedValue) {
+        return records.stream()
+                .map(record -> record.get(seq)
+                        .acceptedValue()
+                        .map(PaxosAcceptorData::getLastAcceptedValue)
+                        .orElseGet(Optional::empty))
+                .filter(optionalPaxosValue ->
+                        optionalPaxosValue.isPresent() && optionalPaxosValue.get().equals(learnedValue))
+                .map(Optional::get)
+                .collect(Collectors.toList());
+    }
+
+    private static Optional<PaxosValue> getGreatestAcceptedValueAtSequence(
+            List<ConsolidatedLearnerAndAcceptorRecord> records, long seq) {
+        return records.stream()
+                .map(record -> record.get(seq)
+                        .acceptedValue()
+                        .map(PaxosAcceptorData::getLastAcceptedValue)
+                        .orElseGet(Optional::empty))
+                .filter(paxosValue -> getPaxosValueData(paxosValue) != null)
+                .map(Optional::get)
+                .max(Comparator.comparingLong(paxosValue ->  PtBytes.toLong(paxosValue.getData())));
+    }
+
+
+    private static byte[] getPaxosValueData(Optional<PaxosValue> learnedValue) {
+        return learnedValue.map(PaxosValue::getData).orElse(null);
```

Comment:
Let's have a comment here that this should only be used for timestamp bound checks: this would behave strangely if we did it on leader election based checks, because there is a semantic difference between `null` and no agreement because paxos phase two didn't happen